packaging_automation/common_tool_methods.py

```diff
@@ -290,3 +323,171 @@ def remove_cloned_code(exec_path: str):
+        return
+    # There could be more than one key with the same name. While statement is used to delete all the private keys
+    # until no key remains (i.e. until exception arises)
+    while True:
+        output = run_with_output(f"gpg --batch --yes --delete-secret-key {key_id}")
+        if output.returncode == 0:
+            print(f"Private key with the name {name} deleted")
+            counter = counter + 1
+        elif output.returncode == 2:  # key not exist
+            print(f"Key for the name {name} does not exist")
+            break
+        else:
+            print(f"Error {output.stderr.decode('ascii')}")
+            break
+        if counter > 0:
+            print(f"{counter} private  keys deleted")
+
+
+def delete_gpg_key_by_name(name: str):
+    delete_private_gpg_key_by_name(name)
+    delete_public_gpg_key_by_name(name)
+
+
+def get_secret_key_by_fingerprint(fingerprint: str) -> str:
+    cmd = f'gpg --batch --export-secret-keys -a "{fingerprint}" | base64'
+    try:
+        # When getting gpg key if gpg key is stored with password and this api which should be used to get keys
+        # without password is called, timeout expired exception is thrown.
+        ps = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, timeout=2)
+
+    except subprocess.TimeoutExpired:
+        raise ValueError(
+            f"Error while getting key. Most probably packaging key is stored with password. "
+            f"Please remove the password when storing key with fingerprint {fingerprint}")
+
+    secret_key = ps.stdout.decode("ascii")
+    return secret_key
+
+
+def get_secret_key_by_fingerprint_with_password(fingerprint: str, passphrase: str) -> str:
+    # When getting gpg key if gpg key is stored with password and if given passphrase is wrong, timeout exception is
+    # thrown.
+    gpg = gnupg.GPG()
+    try:
+        private_key = gpg.export_keys(fingerprint, secret=True, passphrase=passphrase)
+    except subprocess.TimeoutExpired:
+        raise ValueError(
+            f"Error while getting key. Most probably packaging key is stored with password. "
+            f"Please check the password and try again")
+    return base64.b64encode(private_key.encode("ascii")).decode("ascii")
+
+
+def define_rpm_public_key_to_machine(fingerprint: str):
+    run(f"gpg --export -a {fingerprint} >rpm_public.key")
+    run("rpm --import rpm_public.key")
+    os.remove("rpm_public.key")
+
+
+def delete_rpm_key_by_name(key_name: str):
+    result = run_with_output("rpm -q gpg-pubkey --qf %{NAME}-%{VERSION}-%{RELEASE}__%{SUMMARY}-!!")
+    output = result.stdout.decode("ascii")
+    if result.stderr:
+        print(f"Error!!!{output}")
+    else:
+        # Get all the lines which includes rpm keys and get the first item which is rpm key to be used to delete
+        # the key .Example line gpg-pubkey-fd431d51-4ae0493b__gpg(Red Hat, Inc. (release key 2) <security@redhat.com>)
```

Comment:
Can we strictly rely on the fact that line will always have `__` /`!!` as the separator between key and the rest ?
Can you share a link that explicitly declares this format ?

Reply:
Parameterized them as key_separator and line_separator. We are using this statements in the command

Reply:
These parameters are defined by me. It is not a general thing. So there could not be a document about it. __ and !! character series are just seperators

Reply:
![image](https://user-images.githubusercontent.com/1596331/122208691-48711780-ceac-11eb-8270-6b0538873c99.png)

Reply:
I can handle with some escape characters one is required for bash another one is required for python

Reply:
Fixed with the soluyion you shared on Teams